File: camfit-puller/src/camfit_puller/adapters/postgres/review_repo.py
from __future__ import annotations
from typing import Iterable, Iterator, Literal
from ...domain.models import Review
from .pool import PostgresPool


class PostgresReviewReader:
    def __init__(self, pool: PostgresPool):
        self._pool = pool

    def top_for(self, camp_id: str, n: int = 3,
                sort: Literal["score", "recent"] = "score") -> list[Review]:
        order = "score DESC NULLS LAST" if sort == "score" else "review_timestamp DESC"
        with self._pool.conn() as c, c.cursor() as cur:
            cur.execute(
                f"""SELECT id, camp_id, user_nick, season, user_type, num_of_days,
                          score, text, is_clean, is_kind, is_manner, is_convenient, review_timestamp
                    FROM reviews WHERE camp_id=%s ORDER BY {order} LIMIT %s""",
                (camp_id, n),
            )
            return [self._row(r) for r in cur.fetchall()]

    def total_for(self, camp_id: str) -> int:
        with self._pool.conn() as c, c.cursor() as cur:
            cur.execute("SELECT count(*) FROM reviews WHERE camp_id=%s", (camp_id,))
            return cur.fetchone()[0]

    def iter_for(self, camp_id: str) -> Iterator[Review]:
        with self._pool.conn() as c:
            with c.cursor(name=f"rev_{camp_id}") as cur:
                cur.execute(
                    """SELECT id, camp_id, user_nick, season, user_type, num_of_days,
                              score, text, is_clean, is_kind, is_manner, is_convenient, review_timestamp
                       FROM reviews WHERE camp_id=%s""",
                    (camp_id,),
                )
                for r in cur:
                    yield self._row(r)

    @staticmethod
    def _row(r) -> Review:
        return Review(
            id=r[0], camp_id=r[1], user_nick=r[2], season=r[3], user_type=r[4],
            num_of_days=r[5],
            score=float(r[6]) if r[6] is not None else None,
            text=r[7],
            is_clean=r[8], is_kind=r[9], is_manner=r[10], is_convenient=r[11],
            review_timestamp=r[12],
        )


class PostgresReviewWriter:
    def __init__(self, pool: PostgresPool):
        self._pool = pool

    def upsert_many(self, reviews: Iterable[Review]) -> int:
        n = 0
        with self._pool.conn() as c, c.cursor() as cur:
            for r in reviews:
                cur.execute(
                    """INSERT INTO reviews (id, camp_id, user_nick, season, user_type,
                          num_of_days, score, text, is_clean, is_kind, is_manner, is_convenient, review_timestamp)
                       VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                       ON CONFLICT (id) DO UPDATE SET
                         score=EXCLUDED.score, text=EXCLUDED.text""",
                    (r.id, r.camp_id, r.user_nick, r.season, r.user_type, r.num_of_days,
                     r.score, r.text, r.is_clean, r.is_kind, r.is_manner, r.is_convenient,
                     r.review_timestamp),
                )
                cur.execute("DELETE FROM review_medias WHERE review_id=%s", (r.id,))
                if r.medias:
                    cur.executemany(
                        "INSERT INTO review_medias (review_id, idx, url) VALUES (%s,%s,%s)",
                        [(r.id, i, u) for i, u in enumerate(r.medias)],
                    )
                n += 1
        return n

File: backend/be-api-go/internal/adapters/postgres/review_repo.go
// PostgresReviewRepo — pgx port of the read-side methods of
// `adapters.postgres.review_repo.PostgresReviewReader`.
package postgres

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/whyjp/cf/be-api-go/internal/domain"
	"github.com/whyjp/cf/be-api-go/internal/ports"
)

// ReviewRepo implements ports.ReviewReader on top of pgxpool.
type ReviewRepo struct {
	pool *pgxpool.Pool
}

// Compile-time assertion.
var _ ports.ReviewReader = (*ReviewRepo)(nil)

// NewReviewRepo constructs a ReviewRepo from an existing pgxpool.
func NewReviewRepo(pool *pgxpool.Pool) *ReviewRepo {
	return &ReviewRepo{pool: pool}
}

// TopFor returns the top-N reviews for a camp.
//
// Python source: `ORDER BY score DESC NULLS LAST` for sort=score (default),
// `ORDER BY review_timestamp DESC` for sort=recent. medias[] is intentionally
// NOT loaded here — the Python `.top_for` doesn't load it either (review_medias
// is a separate table joined on demand by the writer/iter paths). model_dump
// emits `"medias": []` for the empty default — same as Go.
func (r *ReviewRepo) TopFor(ctx context.Context, campID string, n int, sort string) ([]*domain.Review, error) {
	order := "score DESC NULLS LAST"
	if sort == "recent" {
		order = "review_timestamp DESC"
	}
	sql := `SELECT id, camp_id, user_nick, season, user_type, num_of_days,
	        score, text, is_clean, is_kind, is_manner, is_convenient, review_timestamp
	        FROM reviews WHERE camp_id=$1 ORDER BY ` + order + ` LIMIT $2`
	rows, err := r.pool.Query(ctx, sql, campID, n)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []*domain.Review
	for rows.Next() {
		rev := &domain.Review{Medias: []string{}}
		if err := rows.Scan(
			&rev.ID, &rev.CampID, &rev.UserNick, &rev.Season, &rev.UserType, &rev.NumOfDays,
			&rev.Score, &rev.Text, &rev.IsClean, &rev.IsKind, &rev.IsManner, &rev.IsConvenient,
			&rev.ReviewTimestamp,
		); err != nil {
			return nil, err
		}
		out = append(out, rev)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return out, nil
}

// TotalFor returns the row count of reviews for a camp.
//
// Python source: `SELECT count(*) FROM reviews WHERE camp_id=%s`.
func (r *ReviewRepo) TotalFor(ctx context.Context, campID string) (int, error) {
	var n int
	if err := r.pool.QueryRow(ctx,
		"SELECT count(*) FROM reviews WHERE camp_id=$1", campID).Scan(&n); err != nil {
		return 0, err
	}
	return n, nil
}
